README.md
# NAME

AI::XGBoost - Perl wrapper for XGBoost library [https://github.com/dmlc/xgboost](https://github.com/dmlc/xgboost)

# VERSION

version 0.11

# SYNOPSIS

```perl
use 5.010;
use aliased 'AI::XGBoost::DMatrix';
use AI::XGBoost qw(train);

# We are going to solve a binary classification problem:
#  Mushroom poisonous or not

my $train_data = DMatrix->From(file => 'agaricus.txt.train');
my $test_data = DMatrix->From(file => 'agaricus.txt.test');

# With XGBoost we can solve this problem using 'gbtree' booster
#  and as loss function a logistic regression 'binary:logistic'
#  (Gradient Boosting Regression Tree)
# XGBoost Tree Booster has a lot of parameters that we can tune
# (https://github.com/dmlc/xgboost/blob/master/doc/parameter.md)

my $booster = train(data => $train_data, number_of_rounds => 10, params => {
        objective => 'binary:logistic',
        eta => 1.0,
        max_depth => 2,
        silent => 1
    });

# For binay classification predictions are probability confidence scores in [0, 1]
#  indicating that the label is positive (1 in the first column of agaricus.txt.test)
my $predictions = $booster->predict(data => $test_data);

say join "\n", @$predictions[0 .. 10];

use aliased 'AI::XGBoost::DMatrix';
use AI::XGBoost qw(train);
use Data::Dataset::Classic::Iris;

# We are going to solve a multiple classification problem:
#  determining plant species using a set of flower's measures 

# XGBoost uses number for "class" so we are going to codify classes
my %class = (
    setosa => 0,
    versicolor => 1,
    virginica => 2
);

my $iris = Data::Dataset::Classic::Iris::get();

# Split train and test, label and features
my $train_dataset = [map {$iris->{$_}} grep {$_ ne 'species'} keys %$iris];
my $test_dataset = [map {$iris->{$_}} grep {$_ ne 'species'} keys %$iris];

sub transpose {
# Transposing without using PDL, Data::Table, Data::Frame or other modules
# to keep minimal dependencies
    my $array = shift;
    my @aux = ();
    for my $row (@$array) {
        for my $column (0 .. scalar @$row - 1) {
            push @{$aux[$column]}, $row->[$column];
        }
    }
    return \@aux;
}

$train_dataset = transpose($train_dataset);
$test_dataset = transpose($test_dataset);

my $train_label = [map {$class{$_}} @{$iris->{'species'}}];
my $test_label = [map {$class{$_}} @{$iris->{'species'}}];

my $train_data = DMatrix->From(matrix => $train_dataset, label => $train_label);
my $test_data = DMatrix->From(matrix => $test_dataset, label => $test_label);

# Multiclass problems need a diferent objective function and the number
#  of classes, in this case we are using 'multi:softprob' and
#  num_class => 3
my $booster = train(data => $train_data, number_of_rounds => 20, params => {
        max_depth => 3,
        eta => 0.3,
        silent => 1,
        objective => 'multi:softprob',
        num_class => 3
    });

my $predictions = $booster->predict(data => $test_data);
```

# DESCRIPTION

Perl wrapper for XGBoost library. 

The easiest way to use the wrapper is using `train`, but beforehand 
you need the data to be used contained in a `DMatrix` object

This is a work in progress, feedback, comments, issues, suggestion and
pull requests are welcome!!

XGBoost library is used via [Alien::XGBoost](https://metacpan.org/pod/Alien::XGBoost). That means downloading,
compiling and installing if it's not available in your system.

# FUNCTIONS

## train

Performs gradient boosting using the data and parameters passed

Returns a trained AI::XGBoost::Booster used

### Parameters

- params

    Parameters for the booster object. 

    Full list available: https://github.com/dmlc/xgboost/blob/master/doc/parameter.md 

- data

    AI::XGBoost::DMatrix object used for training

- number\_of\_rounds

    Number of boosting iterations

# ROADMAP

The goal is to make a full wrapper for XGBoost.

## VERSIONS

- 0.2 

    Full C API "easy" to use, with PDL support as [AI::XGBoost::CAPI](https://metacpan.org/pod/AI::XGBoost::CAPI)

    Easy means clients don't have to use [FFI::Platypus](https://metacpan.org/pod/FFI::Platypus) or modules dealing
    with C structures

- 0.25

    Alien package for libxgboost.so/xgboost.dll

- 0.3

    Object oriented API Moose based with DMatrix and Booster classes

- 0.4

    Complete object oriented API

- 0.5

    Use perl signatures ([https://metacpan.org/pod/distribution/perl/pod/perlexperiment.pod#Subroutine-signatures](https://metacpan.org/pod/distribution/perl/pod/perlexperiment.pod#Subroutine-signatures))

# SEE ALSO

- [AI::MXNet](https://metacpan.org/pod/AI::MXNet)
- [FFI::Platypus](https://metacpan.org/pod/FFI::Platypus)
- [NativeCall](https://metacpan.org/pod/NativeCall)

# AUTHOR

Pablo Rodríguez González <pablo.rodriguez.gonzalez@gmail.com>

# COPYRIGHT AND LICENSE

Copyright (c) 2017 by Pablo Rodríguez González.

# CONTRIBUTOR

Ruben <me@ruben.tech>


Changes
0.1     2017-08-16
    * Full raw C API available



examples/iris.pl
use aliased 'AI::XGBoost::DMatrix';
use AI::XGBoost qw(train);
use Data::Dataset::Classic::Iris;

my %class = (
    setosa => 0,
    versicolor => 1,
    virginica => 2
);

my $iris = Data::Dataset::Classic::Iris::get();

my $train_dataset = [map {$iris->{$_}} grep {$_ ne 'species'} keys %$iris];
my $test_dataset = [map {$iris->{$_}} grep {$_ ne 'species'} keys %$iris];
my $train_label = [map {$class{$_}} @{$iris->{'species'}}];
my $test_label = [map {$class{$_}} @{$iris->{'species'}}];

my $train_data = DMatrix->From(matrix => $train_dataset, label => $train_label);
my $test_data = DMatrix->From(matrix => $test_dataset, label => $test_label);

my $booster = train(data => $train_data, number_of_rounds => 20, params => {
        max_depth => 3,
        eta => 0.3,
        silent => 1,
        objective => 'multi:softprob',
        num_class => 3
    });

my $predictions = $booster->predict(data => $test_data);





examples/basic.pl
#!/usr/bin/env perl 

use strict;
use warnings;
use utf8;
use 5.010;
use aliased 'AI::XGBoost::DMatrix';
use AI::XGBoost qw(train);

# We are going to solve a binary classification problem:
#  Mushroom poisonous or not

my $train_data = DMatrix->FromFile(filename => 'agaricus.txt.train');
my $test_data = DMatrix->FromFile(filename => 'agaricus.txt.test');

# With XGBoost we can solve this problem using 'gbtree' booster
#  and as loss function a logistic regression 'binary:logistic'
#  (Gradient Boosting Regression Tree)
# XGBoost Tree Booster has some parameters that we can tune
# For this example we are going to use some "defaults"

say "Training...";
my $booster = train(data => $train_data, number_of_rounds => 10);

say "Predicting...";
my $predictions = $booster->predict(data => $test_data);

say join "\n", @$predictions[0 .. 10];








examples/capi_raw.pl
use 5.010;
use AI::XGBoost::CAPI::RAW;
use FFI::Platypus;

my $silent = 0;
my ($dtrain, $dtest) = (0, 0);

AI::XGBoost::CAPI::RAW::XGDMatrixCreateFromFile('agaricus.txt.test', $silent, \$dtest);
AI::XGBoost::CAPI::RAW::XGDMatrixCreateFromFile('agaricus.txt.train', $silent, \$dtrain);

my ($rows, $cols) = (0, 0);
AI::XGBoost::CAPI::RAW::XGDMatrixNumRow($dtrain, \$rows);
AI::XGBoost::CAPI::RAW::XGDMatrixNumCol($dtrain, \$cols);
say "Dimensions: $rows, $cols";

my $booster = 0;

AI::XGBoost::CAPI::RAW::XGBoosterCreate( [$dtrain] , 1, \$booster);

for my $iter (0 .. 10) {
    AI::XGBoost::CAPI::RAW::XGBoosterUpdateOneIter($booster, $iter, $dtrain);
}

my $out_len = 0;
my $out_result = 0;

AI::XGBoost::CAPI::RAW::XGBoosterPredict($booster, $dtest, 0, 0, \$out_len, \$out_result);
my $ffi = FFI::Platypus->new();
my $predictions = $ffi->cast(opaque => "float[$out_len]", $out_result);

#say join "\n", @$predictions;

AI::XGBoost::CAPI::RAW::XGBoosterFree($booster);






lib/AI/XGBoost/CAPI.pm
package AI::XGBoost::CAPI;
use strict;
use warnings;

use Exporter::Easy (
        TAGS => [
            all => [qw(
                XGDMatrixCreateFromFile
                XGDMatrixCreateFromMat
                XGDMatrixNumRow
                XGDMatrixNumCol
                XGDMatrixSetFloatInfo
                XGDMatrixGetFloatInfo
                XGDMatrixSetUintInfo
                XGDMatrixGetUintInfo
                XGDMatrixSaveBinary
                XGDMatrixSliceDMatrix
                XGDMatrixFree
                XGBoosterCreate
                XGBoosterSetParam
                XGBoosterSetAttr
                XGBoosterGetAttr
                XGBoosterGetAttrNames
                XGBoosterUpdateOneIter
                XGBoosterBoostOneIter
                XGBoosterPredict
                XGBoosterFree
                XGBoosterDumpModel
                XGBoosterDumpModelEx
                XGBoosterDumpModelWithFeatures
                XGBoosterDumpModelExWithFeatures
                )]
        ]
    );
use AI::XGBoost::CAPI::RAW;
use FFI::Platypus;
use Exception::Class (
    'XGBoostException'
);

# VERSION

# ABSTRACT: Perl wrapper for XGBoost C API https://github.com/dmlc/xgboost

=encoding utf-8

=head1 SYNOPSIS

# EXAMPLE: examples/capi.pl

=head1 DESCRIPTION

Perlified wrapper for the C API

=head2 Error handling

XGBoost c api functions returns some int to signal the presence/absence of error.
In this module that is achieved using Exceptions from L<Exception::Class>

=head1 FUNCTIONS

=cut

=head2 XGDMatrixCreateFromFile

Load a data matrix

Parameters:

=over 4

=item filename

the name of the file

=item silent 

whether print messages during loading

=back

Returns a loaded data matrix

=cut 

sub XGDMatrixCreateFromFile {
    my ($filename, $silent) = @_;
    $silent //= 1;
    my $matrix = 0;
    my $error = AI::XGBoost::CAPI::RAW::XGDMatrixCreateFromFile($filename, $silent, \$matrix);
    _CheckCall($error);
    return $matrix;
}

=head2 XGDMatrixCreateFromMat

Create from dense matrix

Parameters:

=over 4

=item matrix

matrix data

=item missing

value indicating missing data (optional)

=back

Returns a loaded data matrix

=cut 

sub XGDMatrixCreateFromMat {
    my ($data, $missing) = @_;
    $missing //= "NaN";
    # TODO Support simple arrays
    # TODO Support PDL
    # TODO ¿Adapters?
    my $data_adapter = [ map {@$_} @$data ];
    my $nrows = scalar @$data;
    my $ncols = scalar @{$data->[0]};
    my $matrix = 0;
    my $error = AI::XGBoost::CAPI::RAW::XGDMatrixCreateFromMat($data_adapter, $nrows, $ncols, $missing, \$matrix);
    _CheckCall($error);
    return $matrix;
}

=head2 XGDMatrixNumRow

Get number of rows

Parameters:

=over 4

=item matrix

DMatrix

=back

=cut

sub XGDMatrixNumRow {
    my ($matrix) = @_;
    my $rows = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGDMatrixNumRow($matrix, \$rows) );
    return $rows;
}

=head2 XGDMatrixNumCol

Get number of cols

Parameters:

=over 4

=item matrix

DMatrix

=back

=cut

sub XGDMatrixNumCol {
    my ($matrix) = @_;
    my $cols = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGDMatrixNumCol($matrix, \$cols) );
    return $cols;
}

=head2 XGDMatrixSetFloatInfo 

=cut

sub XGDMatrixSetFloatInfo {
    my ($matrix, $info, $data) = @_;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGDMatrixSetFloatInfo($matrix, $info, $data, scalar @$data) );
}

=head2 XGDMatrixGetFloatInfo 

=cut

sub XGDMatrixGetFloatInfo {
    my ($matrix, $info) = @_;
    my $out_len = 0;
    my $out_result = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGDMatrixGetFloatInfo($matrix, $info, \$out_len, \$out_result) );
    my $ffi = FFI::Platypus->new();
    return $ffi->cast(opaque => "float[$out_len]", $out_result);
}

=head2 XGDMatrixSetUintInfo 

=cut

sub XGDMatrixSetUintInfo {
    my ($matrix, $info, $data) = @_;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGDMatrixSetUintInfo($matrix, $info, $data, scalar @$data) );
}

=head2 XGDMatrixGetUintInfo 

=cut

sub XGDMatrixGetUintInfo {
    my ($matrix, $info) = @_;
    my $out_len = 0;
    my $out_result = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGDMatrixGetUintInfo($matrix, $info, \$out_len, \$out_result) );
    my $ffi = FFI::Platypus->new();
    return $ffi->cast(opaque => "uint32[$out_len]", $out_result);
}

=head2 XGDMatrixSaveBinary

=cut

sub XGDMatrixSaveBinary {
    my ($matrix, $filename, $silent) = @_;
    $silent //= 1;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGDMatrixSaveBinary($matrix, $filename, $silent));
}

=head2 XGDMatrixSliceDMatrix

=cut

sub XGDMatrixSliceDMatrix {
    my ($matrix, $list_of_indices) = @_;
    my $new_matrix = 0;
    my $error = AI::XGBoost::CAPI::RAW::XGDMatrixSliceDMatrix($matrix, $list_of_indices, scalar @$list_of_indices, \$new_matrix);
    _CheckCall($error);
    return $new_matrix;
}

=head2 XGDMatrixFree

Free space in data matrix

Parameters:

=over 4

=item matrix

DMatrix to be freed

=back

=cut

sub XGDMatrixFree {
    my ($matrix) = @_;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGDMatrixFree($matrix) );
    return (); 
}

=head2 XGBoosterCreate

Create XGBoost learner

Parameters:

=over 4

=item matrices

matrices that are set to be cached

=back

=cut

sub XGBoosterCreate {
    my ($matrices) = @_;
    my $booster = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterCreate($matrices, scalar @$matrices, \$booster) );
    return $booster;
}

=head2 XGBoosterSetParam

=cut

sub XGBoosterSetParam {
    my ($booster, $name, $value) = @_;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterSetParam($booster, $name, $value) );
}

=head2 XGBoosterSetAttr

=cut

sub XGBoosterSetAttr {
    my ($booster, $name, $value) = @_;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterSetAttr($booster, $name, $value) );
}

=head2 XGBoosterGetAttr

=cut

sub XGBoosterGetAttr {
    my ($booster, $name) = @_;
    my $value = 0;
    my $success = -1;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterGetAttr($booster, $name, \$value, \$success) );
    if ($success) {
        my $ffi = FFI::Platypus->new();
        return $ffi->cast(opaque => "string", $value); 
    }
    return ();
}

=head2 XGBoosterGetAttrNames

=cut

sub XGBoosterGetAttrNames {
    my ($booster) = @_;
    my $out_len = 0;
    my $out_result = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterGetAttrNames($booster, \$out_len, \$out_result) );
    my $ffi = FFI::Platypus->new();
    $out_result = $ffi->cast(opaque => "opaque[$out_len]", $out_result);
    return [map {$ffi->cast(opaque => "string", $_)} @$out_result ];
}

=head2 XGBoosterUpdateOneIter

Update the model in one round using train matrix

Parameters:

=over 4

=item booster

XGBoost learner to train

=item iter

current iteration rounds

=item train_matrix

training data

=back

=cut

sub XGBoosterUpdateOneIter {
    my ($booster, $iter, $train_matrix) = @_;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterUpdateOneIter($booster, $iter, $train_matrix) );
    return ();
}

=head2 XGBoosterBoostOneIter

=cut

sub XGBoosterBoostOneIter {
    my ($booster, $train_matrix, $gradient, $hessian) = @_;
    my $out_result = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterBoostOneIter($booster, $train_matrix, $gradient, $hessian, scalar (@$gradient)));
    return ();
}

=head2 XGBoosterEvalOneIter

=cut

sub XGBoosterEvalOneIter {
    my ($booster, $iter, $matrices, $matrices_names) = @_;
    my $out_result = 0;
    my $number_of_matrices = scalar @$matrices;
    my $ffi = FFI::Platypus->new();
    my $array_of_opaque_matrices_names = [map {$ffi->cast(string => "opaque", $_)} @$matrices_names];
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterEvalOneIter($booster, $iter, $matrices, $array_of_opaque_matrices_names, $number_of_matrices, \$out_result));
    $out_result = $ffi->cast(opaque => "opaque[$number_of_matrices]", $out_result);
    return [map {$ffi->cast(opaque => "string", $_)} @$out_result ];
}

=head2 XGBoosterPredict

Make prediction based on train matrix

Parameters:

=over 4

=item booster

XGBoost learner 

=item data_matrix

Data matrix with the elements to predict

=item option_mask

bit-mask of options taken in prediction, possible values

=over 4

=item 

0: normal prediction

=item

1: output margin instead of transformed value

=item

2: output leaf index of trees instead of leaf value, note leaf index is unique per tree

=item 

4: output feature contributions to individual predictions

=back


=item ntree_limit

limit number of trees used for prediction, this is only valid for boosted trees
when the parameter is set to 0, we will use all the trees

=back

Returns an arrayref with the predictions corresponding to the rows of data matrix

=cut 

sub XGBoosterPredict {
    my ($booster, $data_matrix, $option_mask, $ntree_limit) = @_;
    $option_mask //= 0;
    $ntree_limit //= 0;
    my $out_len = 0;
    my $out_result = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterPredict($booster, $data_matrix, $option_mask, $ntree_limit, \$out_len, \$out_result) );
    my $ffi = FFI::Platypus->new();
    return $ffi->cast(opaque => "float[$out_len]", $out_result);
}


=head2 XGBoosterDumpModel

=cut

sub XGBoosterDumpModel {
    my ($booster, $feature_map, $with_stats) = @_;
    $feature_map //= "";
    $with_stats //= 1;
    my $out_len = 0;
    my $out_result = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterDumpModel($booster, $feature_map, $with_stats, \$out_len, \$out_result) );
    my $ffi = FFI::Platypus->new();
    $out_result = $ffi->cast(opaque => "opaque[$out_len]", $out_result);
    return [map {$ffi->cast(opaque => "string", $_)} @$out_result ];
}

=head2 XGBoosterDumpModelEx

=cut

sub XGBoosterDumpModelEx {
    my ($booster, $feature_map, $with_stats, $format) = @_;
    $feature_map //= "";
    $with_stats //= 1;
    my $out_len = 0;
    my $out_result = 0;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterDumpModelEx($booster, $feature_map, $with_stats, $format, \$out_len, \$out_result) );
    my $ffi = FFI::Platypus->new();
    $out_result = $ffi->cast(opaque => "opaque[$out_len]", $out_result);
    return [map {$ffi->cast(opaque => "string", $_)} @$out_result ];
}

=head2 XGBoosterDumpModelWithFeatures

=cut

sub XGBoosterDumpModelWithFeatures {
    my ($booster, $feature_names, $feature_types, $with_stats) = @_;
    $with_stats //= 1;
    my $out_len = 0;
    my $out_result = 0;
    my $ffi = FFI::Platypus->new();
    my $number_of_features = scalar @$feature_names;
    my $array_of_opaque_feature_names = [map {$ffi->cast(string => "opaque", $_)} @$feature_names];
    my $array_of_opaque_feature_types = [map {$ffi->cast(string => "opaque", $_)} @$feature_types];
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterDumpModelWithFeatures($booster, $number_of_features, $array_of_opaque_feature_names, $array_of_opaque_feature_types, $with_stats, \$out_len, \$out_result) );
    $out_result = $ffi->cast(opaque => "opaque[$out_len]", $out_result);
    return [map {$ffi->cast(opaque => "string", $_)} @$out_result ];
}

=head2 XGBoosterDumpModelExWithFeatures

=cut

sub XGBoosterDumpModelExWithFeatures {
    my ($booster, $feature_names, $feature_types, $with_stats, $format) = @_;
    my $out_len = 0;
    my $out_result = 0;
    my $ffi = FFI::Platypus->new();
    my $number_of_features = scalar @$feature_names;
    my $array_of_opaque_feature_names = [map {$ffi->cast(string => "opaque", $_)} @$feature_names];
    my $array_of_opaque_feature_types = [map {$ffi->cast(string => "opaque", $_)} @$feature_types];
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterDumpModelExWithFeatures($booster, $number_of_features, $array_of_opaque_feature_names, $array_of_opaque_feature_types, $with_stats, $format, \$out_len, \$out_result) );
    $out_result = $ffi->cast(opaque => "opaque[$out_len]", $out_result);
    return [map {$ffi->cast(opaque => "string", $_)} @$out_result ];
}

=head2 XGBoosterFree

Free booster object

Parameters:

=over 4

=item booster

booster to be freed

=back

=cut

sub XGBoosterFree {
    my ($booster) = @_;
    _CheckCall( AI::XGBoost::CAPI::RAW::XGBoosterFree($booster) );
    return ();
}

# _CheckCall
# 
#  Check return code and if necesary, launch an exception
#
sub _CheckCall {
    my ($return_code) = @_;
    if ($return_code) {
        my $error_message = AI::XGBoost::CAPI::RAW::XGBGetLastError();
        XGBoostException->throw(error => $error_message);
    }
}

1;


lib/AI/XGBoost/Booster.pm
package AI::XGBoost::Booster;

use strict;
use warnings;
use utf8;

# VERSION

# ABSTRACT: XGBoost main class for training, prediction and evaluation

use Moose;
use AI::XGBoost::CAPI qw(:all);
use namespace::autoclean;

=encoding utf-8

=head1 SYNOPSIS

# EXAMPLE: examples/basic.pl

=head1 DESCRIPTION

Booster objects control training, prediction and evaluation

Work In Progress, the API may change. Comments and suggestions are welcome!

=head1 METHODS

=cut

has _handle => (
    is => 'rw',
    init_arg => undef,
);

=head2 update

Update one iteration

=head3 Parameters

=over 4

=item iteration

Current iteration number

=item dtrain

Training data (AI::XGBoost::DMatrix)

=back

=cut

sub update {
    my $self = shift;
    my %args = @_;
    my ($iteration, $dtrain) = @args{qw(iteration dtrain)};
    XGBoosterUpdateOneIter($self->_handle, $iteration, $dtrain->handle);
    return $self;
}

=head2 boost

Boost one iteration using your own gradient

=head3 Parameters

=over 4

=item dtrain

Training data (AI::XGBoost::DMatrix)

=item grad

Gradient of your objective function (Reference to an array)

=item hess

Hessian of your objective function, that is, second order gradient (Reference to an array)

=back

=cut

sub boost {
    my $self = shift;
    my %args = @_;
    my ($dtrain, $grad, $hess) = @args{qw(dtrain grad hess)};
    XGBoosterBoostOneIter($self->_handle, $dtrain, $grad, $hess);
    return $self;
}

=head2 predict

Predict data using the trained model

=head3 Parameters

=over 4

=item data

Data to predict

=back

=cut

sub predict {
    my $self = shift;
    my %args = @_;
    my $data = $args{'data'};
    my $result = XGBoosterPredict($self->_handle, $data->handle);
    my $result_size = scalar @$result;
    my $matrix_rows = $data->num_row;
    if ($result_size != $matrix_rows && $result_size % $matrix_rows == 0) {
        my $col_size = $result_size / $matrix_rows;
        return [map {[@$result[$_ * $col_size .. $_ * $col_size + $col_size - 1]]} 0 .. $matrix_rows - 1];
    }
    return $result;
}

=head2 set_param

Set booster parameter

=head3 Example

    $booster->set_param('objective', 'binary:logistic');

=cut

sub set_param {
    my $self = shift;
    my ($name, $value) = @_;
    XGBoosterSetParam($self->_handle, $name, $value);
    return $self;
}

=head2 set_attr

Set a string attribute

=cut

sub set_attr {
    my $self = shift;
    my ($name, $value) = @_;
    XGBoosterSetAttr($self->_handle, $name, $value);
    return $self;
}

=head2 get_attr

Get a string attribute

=cut

sub get_attr {
    my $self = shift;
    my ($name) = @_;
    XGBoosterGetAttr($self->_handle, $name);
}

=head2 get_score

Get importance of each feature

=head3 Parameters

=over 4

=item importance_type

Type of importance. Valid values:

=over 4

=item weight

Number of times a feature is used to split the data across all trees

=item gain

Average gain of the feature when it is used in trees

=item cover

Average coverage of the feature when it is used in trees

=back

=item fmap

Name of feature map file

=back

=cut

sub get_score {
    my $self = shift;
    my %args = @_;
    my ($fmap, $importance_type) = @args{qw(fmap importance_type)};

    if ($importance_type eq "weight") {
        my @trees = $self->get_dump;
    } else {

    }

}


=head2 get_dump

=cut

sub get_dump {
    my $self = shift;
    return XGBoosterDumpModelEx($self->_handle, "", 1, "text");
}


=head2 attributes

Returns all attributes of the booster as a HASHREF

=cut

sub attributes {
    my $self = shift;
    return {map {$_ => $self->get_attr($_)} @{XGBoosterGetAttrNames($self->_handle)}};
}

=head2 TO_JSON

Serialize the booster to JSON.

This method is to be used with the option C<convert_blessed> from L<JSON>.
(See L<https://metacpan.org/pod/JSON#OBJECT-SERIALISATION>)

Warning: this API is subject to changes

=cut

sub TO_JSON {
    my $self = shift;
    my $trees = XGBoosterDumpModelEx($self->_handle, "", 1, "json");
    return "[" . join (',', @$trees) . "]";
}

=head2 BUILD

Use new, this method is just an internal helper

=cut

sub BUILD {
    my $self = shift;
    my $args = shift;
    $self->_handle(XGBoosterCreate([map {$_->handle } @{$args->{'cache'}}]));
}

=head2 DEMOLISH

Internal destructor. This method is called automatically

=cut

sub DEMOLISH {
    my $self = shift();
    XGBoosterFree($self->_handle);
}

__PACKAGE__->meta->make_immutable();

1;
 



lib/AI/XGBoost.pm
package AI::XGBoost;
use strict;
use warnings;

use AI::XGBoost::Booster;
use Exporter::Easy ( OK => ['train'] );

# VERSION

# ABSTRACT: Perl wrapper for XGBoost library L<https://github.com/dmlc/xgboost>

=encoding utf-8

=head1 SYNOPSIS

# EXAMPLE: examples/basic.pl

# EXAMPLE: examples/iris.pl

=head1 DESCRIPTION

Perl wrapper for XGBoost library. 

The easiest way to use the wrapper is using C<train>, but beforehand 
you need the data to be used contained in a C<DMatrix> object

This is a work in progress, feedback, comments, issues, suggestion and
pull requests are welcome!!

XGBoost library is used via L<Alien::XGBoost>. That means downloading,
compiling and installing if it's not available in your system.

=head1 FUNCTIONS

=cut

=head2 train

Performs gradient boosting using the data and parameters passed

Returns a trained AI::XGBoost::Booster used

=head3 Parameters

=over 4

=item params

Parameters for the booster object. 

Full list available: https://github.com/dmlc/xgboost/blob/master/doc/parameter.md 

=item data

AI::XGBoost::DMatrix object used for training

=item number_of_rounds

Number of boosting iterations

=back

=cut

sub train {
    my %args = @_;
    my ( $params, $data, $number_of_rounds ) = @args{qw(params data number_of_rounds)};

    my $booster = AI::XGBoost::Booster->new( cache => [$data] );
    if ( defined $params ) {
        while ( my ( $name, $value ) = each %$params ) {
            $booster->set_param( $name, $value );
        }
    }
    for my $iteration ( 0 .. $number_of_rounds - 1 ) {
        $booster->update( dtrain => $data, iteration => $iteration );
    }
    return $booster;
}

=head1 ROADMAP

The goal is to make a full wrapper for XGBoost.

=head2 VERSIONS

=over 4

=item 0.2 

Full C API "easy" to use, with PDL support as L<AI::XGBoost::CAPI>

Easy means clients don't have to use L<FFI::Platypus> or modules dealing
with C structures

=item 0.25

Alien package for libxgboost.so/xgboost.dll

=item 0.3

Object oriented API Moose based with DMatrix and Booster classes

=item 0.4

Complete object oriented API

=item 0.5

Use perl signatures (L<https://metacpan.org/pod/distribution/perl/pod/perlexperiment.pod#Subroutine-signatures>)

=back

=head1 SEE ALSO

=over 4

=item L<AI::MXNet>

=item L<FFI::Platypus>

=item L<NativeCall>

=back

=cut

1;
